### ISSUE TITLE
Resizing icon media containing ico files not working

### ISSUE THREAD

**Issue by Gildum** [azuravian/playnite-gamemediaprocessor#6]
Resizing Icon media that have the "ico" file extension doesn't properly work. I think there is no easy solution to this as ico files usually consist of multiple seperate images with different dimensions.

**example**
- the ico file I want to resize consists of 4 seperate images with different dimensions: 128x128, 64x64, 32x32, 16x16
- for the target dimension I choose: 50x50
- result: the icon doesn't get processed
- now I choose another target dimension, which is smaller than the smallest image file of the icon: 12x12
- result: icon gets processed with the result that all 4 images of the icon are resized to 12x12

**How it could work instead**
- rather than getting resized, the ico files instead get their seperate images removed, depending on the target dimension
- e.g. I choose the target: 50x50
- result: the icon from the example now only consists of seperate 2 images: 32x32 & 16x16
- so basically any additional images of the ico files that have a higher resolution than the target dimension get removed, thus reducing file size

Some ico files have a file size of ~500KB, so resizing them would be quite useful. But as mentioned, this is probably not easy to implement.

On a sidenote, there is an icon edit tool "IcoFX" (last freeware version is 1.64) that has a batch process function, where such a feature is possible (reducing file size of multiple ico files at once).

**Comment by azuravian** [comment]
I don't want to require an additional tool, but let me see if I can work out a way to implement your method for ICO files.

**Comment by Gildum** [comment]
I recently found the proper imagemagick command for resizing ico files. It takes the largest image of the input ico file as the base (e.g. 128x128 if ico file consists of 128x128, 32x32, 16x16 images) and creates a new ico file with specific dimensions. The command is:

`magick in.ico ( -clone 0--1 -layers Merge ) -channel A -evaluate Multiply %[fx:w==u[-1].w?1:0] +channel +delete -background None -layers merge -define icon:auto-resize="48,32,16" out.ico`

**source:**
https://legacy.imagemagick.org/discourse-server/viewtopic.php?p=164113#p164113
https://legacy.imagemagick.org/discourse-server~/viewtopic73b3.html?p=133106#p133106

The command `-define icon:auto-resize=""` allows you to set any dimensions for the new ico file. So in the example above, the output ico would consist if 3 images of the sizes 48x48, 32x32, and 16x16. 

**Some notes**
- 16x16 is the lowest possible resolution; smaller values will result in corrupted ico files
- In playnite's default theme the default icon height in details view is 26 (adjustable) and 20 in list view; that's why I think that for resizing ico files, a target of 48x48 should be good enough
- the ico display behaviour in Playnite (Desktop mode) is inconsistent. In some instances, playnite will use the lowest resolution image of the ico file, resulting in very blurry icons. In my testing, it happend with the following ico dimension combinations:
	- 32,24,16 (while 48,24,16 was fine)
	- 48,26 (while 48,16 was fine)
- it seems the safest way to ensure ico files are properly scaled in playnite is to just use single-image ico files (e.g. just 48x48). 
- a 48x48 ico file is just 10kB. For those large ico files of 550kB that's a big file size reduction of 98%
- 2 out of ~100 icons I resized with this method resulted in glitchy icons (artifacts). Original files and output results: [ico.zip](https://github.com/azuravian/playnite-gamemediaprocessor/files/8707150/ico.zip)

**Comment by azuravian** [comment]
Would you be interested in also converting from another format (such as png) to ico as part of the process?  What I may do, is add a separate option for either selected or all icons that prompts for size and just outputs icons in the requested size.  Let me know what you think.

**Comment by Gildum** [comment]
Yes I think it's best to process all icon files at once, regardless of their actual format (from what I have seen they are mainly .ico and .png). Your suggestion sounds great and straightforward. As an example for the user input of the output ico size, it could be: `48,32,16` or just a single number for a single image ico: `64`

The ImageMagick command I have seen suggested in forum posts to convert single image files (.png, .jpg) to .ico is:
`magick in.ico -resize 256x256 -gravity center -background none -extent 256x256 -define icon:auto-resize="48,32,16" out.ico`

**Comment by azuravian** [comment]
Sorry it took so long for me to ever get back to this, but this has been fixed in 1.9.5, as part of commit 29b0864.
